pam/internal/gdm/conversation.go

```diff
@@ -159,3 +161,37 @@ func EmitEvent(pamMTx pam.ModuleTransaction, event Event) error {
 
 	return nil
 }
+
+// DataConversationFunc is an adapter to allow the use of ordinary
+// functions as gdm conversation callbacks.
+type DataConversationFunc func(*Data) (*Data, error)
+
+// RespondPAMBinary is a conversation callback adapter.
+func (f DataConversationFunc) RespondPAMBinary(ptr pam.BinaryPointer) (pam.BinaryPointer, error) {
+	json, err := decodeJSONProtoMessage(ptr)
+	if err != nil {
+		return nil, err
+	}
+	gdmData, err := NewDataFromJSON(json)
+	if err != nil {
+		return nil, err
+	}
+	retData, err := f(gdmData)
+	if err != nil {
+		return nil, err
+	}
+	json, err = retData.JSON()
+	if err != nil {
+		return nil, err
+	}
+	msg, err := newJSONProtoMessage(json)
+	if err != nil {
+		return nil, err
+	}
+	return pam.BinaryPointer(msg), nil
+}
```

Comment:
A function with a method? I'm pretty sure this is not ideal. You don't rely on anything from `f` besides the actual `f(gdmData)` call. IMHO, the function should be an argument of `RespondPAMBinary`.